R/download.R

```diff
@@ -4,6 +4,75 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
+#' Download Bugzilla issues using the Bugzilla REST API
+#'
+#' Downloads bugzilla issues and returns a list of bugzilla bug ids for use with
+#'  \code{\link{download_bugzilla_comments_from_rest_api}}
+#'
+#' @param bugzilla_site URL to specific bugzilla site
+#' @param start_timestamp when to start bug retrieval (ex. 2023-01-01T00:14:57Z)
+#' @param save_folder_path the full *folder* path where the bugzilla issues will be stored
+#' @seealso \code{\link{download_bugzilla_comments_from_rest_api}}
+#' @export
+download_bugzilla_issues_from_rest_api <- function(bugzilla_site, start_timestamp, save_folder_path){
+  # Format link to retrieve data using Bugzilla REST API
+  bugzilla_site <- paste(bugzilla_site, "/rest", sep="")
+
+  # Make sure folder path is correctly formatted
+  if (stringr::str_sub(save_folder_path,-1) != "/"){
+    save_folder_path <-paste0(save_folder_path, "/")
+  }
+
+  # Get the issues starting from the specified date
+  issues <- httr::GET(paste(bugzilla_site, "/bug", "?creation_time=", start_timestamp, sep=""),
+                      limit=20, offset=0,
+                      httr::write_disk(file.path(paste0(save_folder_path, 0, ".json")), overwrite = TRUE))
+
+  bug_ids = c()
+  offset = 20
+  page = 0
+  # Get the bugs and save each page to a json file in the specified save_folder_path
+  while(httr::status_code(issues) == 200){
+    issues = httr::GET(paste(bugzilla_site, "/bug", "?creation_time=", start_timestamp, sep=""),
+                       limit=20, offset=offset,
+                       httr::write_disk(file.path(paste0(save_folder_path, page, ".json")), overwrite = TRUE))
```

Comment:
This is very good!

Reply:
Also, you may want to see what Perceval does with their restful API endpoint. Do they ask a parameter from the user? Surely they had to also decide on this question!

Reply:
Hi Carlos,
I was looking at https://github.com/sailuh/kaiaulu/blob/716a46e546d5293f3b2856cace62311c95aae977/R/github.R#L329-L350

I didn't see an exact answer to using a different condition for the while loop, but to solve the problem where I end up writing an empty json file at the end of my download issues function, I was wondering if I could use write_json instead of htrr:write_disk? That way, I can separate the GET request from actually writing the json to a file. I tried it out and I can do it so that I don't generate an empty file at the end.


```r
  while(length(httr::content(issues)$bugs) > 0){
    issues <- httr::GET(paste(bugzilla_site, "/bug", "?creation_time=", start_timestamp,
                             "&limit=20", "&offset=", offset, sep=""))

    if (length(httr::content(issues)$bugs) == 0){
      print("Break so we don't write empty file to folder.")
      break
    }

    # Writing the file to a json, separate from the GET request
    jsonlite::write_json(httr::content(issues), paste0(save_folder_path, page, ".json"), pretty=TRUE, auto_unbox=TRUE)

    json_issues <- rawToChar(httr::content(issues, as = "raw"))
    current_bug_ids <- jsonlite::fromJSON(json_issues)[["bugs"]][["id"]]
    bug_ids <- c(bug_ids, current_bug_ids)
    offset <- offset + 20
    page <- page + 1
  }
```

Reply:
>Interesting. Did you check the API documentation to see if there isn't a command where you can pull the limit set by the administrator? Also could you hyperlink where you got this from?

The link was in my comment https://bugzilla.readthedocs.io/en/latest/api/core/v1/bug.html#search-bugs. The "limit" is in the Search Bugs section table where it lists the parameters. I have not yet found anything in the API documentation to pull limit set by the administrator.

Also, you may want to see what Perceval does with their restful API endpoint. Do they ask a parameter from the user? Surely they had to also decide on this question!

It seems like Perceval is getting a max_bugs parameter 
https://github.com/chaoss/grimoirelab-perceval/blob/fd6dde67a2ce5902735d849fae7f6f68d6d85de3/perceval/backends/core/bugzillarest.py#L76-L78

And using that as the limit
https://github.com/chaoss/grimoirelab-perceval/blob/fd6dde67a2ce5902735d849fae7f6f68d6d85de3/perceval/backends/core/bugzillarest.py#L189-L190

It's an option you can use in the perceval bugzillarest command.

This appears in the -h section for perceval bugzillarest in the terminal:
```
Bugzilla REST arguments:
  --max-bugs MAX_BUGS   Maximum number of bugs requested on the same query
```

Reply:
> Sounds good! Also, did you see my response https://github.com/sailuh/kaiaulu/pull/177#discussion_r1162520518? 

> I didn't see an exact answer to using a different condition for the while loop, but to solve the problem where I end up writing an empty json file at the end of my download issues function, I was wondering if I could use write_json instead of htrr:write_disk? That way, I can separate the GET request from actually writing the json to a file. I tried it out and I can do it so that I don't generate an empty file at the end.

This is it: 

https://github.com/sailuh/kaiaulu/blob/716a46e546d5293f3b2856cace62311c95aae977/R/github.R#L336

More specifically the `!is.null(gh_response)` 

With GitHub API, we request a new page with this:

https://github.com/sailuh/kaiaulu/blob/716a46e546d5293f3b2856cace62311c95aae977/R/github.R#L344

The function used, in turn uses a library that wraps around GitHub API:

https://github.com/sailuh/kaiaulu/blob/716a46e546d5293f3b2856cace62311c95aae977/R/github.R#L290

In essence, in your code, you request the next page at every step of the loop, incrementally somewhere. Surely, at some point you will request a page + 1 that doesn't exist. You have to examine what httr returns to you when this happens, and test for it on the while loop. 

I think that's what you were trying to do with status code, but you said it won't stop the condition. Why? try running the function on a page that doesn't exist and one that exists, or maybe look how that function of gh_response is able to return a NULL value. This is likely the code logic you are missing, unless I am misunderstanding the issue.